src/popls/bp.rs

```diff
@@ -278,31 +278,37 @@ impl OpRequest<Option<WoutAssignment>> {
     }
 }
 
-impl<T> OpRequest<T> {
-    pub fn check<C: StateCalc>(&self, mut calc: C) -> Result<C, UnmatchedState> {
-        for inp in &self.using {
+impl<S: Supply, P: Pile> Stockpile<S, P> {
+    pub fn check_request<T>(&mut self, request: &OpRequest<T>) -> Result<(), UnmatchedState> {
+        let state = self.state();
+        let api = &self.stock().articles().schema.default_api;
+        let mut calcs = BTreeMap::new();
+        for inp in &request.using {
+            let state_name = state
+                .owned
+                .iter()
+                .find_map(|(state_name, map)| {
+                    map.keys()
+                        .find(|addr| **addr == inp.addr)
+                        .map(|_| state_name)
+                })
+                .expect("unknown state included in the stock");
+            let calc = calcs
+                .entry(state_name)
+                .or_insert_with(|| api.calculate(state_name.clone()));
             calc.accumulate(&inp.val)?;
         }
-        for out in &self.owned {
+        for out in &request.owned {
+            let calc = calcs
+                .entry(&out.name)
+                .or_insert_with(|| api.calculate(out.name.clone()));
             calc.lessen(&out.state.data)?;
         }
-        if !calc.diff()?.is_empty() {
-            Err(UnmatchedState::NotEnoughChange)
-        } else {
-            Ok(calc)
+        for (state_name, calc) in calcs {
+            if !calc.diff()?.is_empty() {
```

Comment:
I believe that verification fails if the contract is fungible.

According to the current [implementation](https://github.com/AluVM/sonic/blob/v0.12.0-beta.4/api/src/adaptors/embedded.rs#L280-L284), even if the sum of the outputs minus the inputs was zero, the diff method will return a vector that contains one element.